File: Problem_2_bruteforce.cpp
//Space Complexity: O(1)
//Time Complexity: O(2^n), where n is the length of the array.
//Time Limit exceeded  on leetcode.

class Solution {

private:
    int helper(vector<int>& nums, int index, int amount){
        //base case
        if(index >= nums.size()) return amount;
        //logic
        //choose
        int case1 = helper(nums, index+2, amount+nums.at(index));
        //not choose
        int case2 = helper(nums, index+1, amount);
        return max(case1, case2);
        
    }
    

public:
    int rob(vector<int>& nums) {
        if(nums.size()==0) return 0;
        return helper(nums,0,0);
    }
};

File: Problem2_bruteforce.py
#
# Time Complexity = O(2^n)
# Space Complexity = O(n)
# Works on leetcode
#

class Solution:
    def rob(self, nums: List[int]) -> int:
        def helper(index):
            if index >= len(nums):
                return 0
            if index == len(nums)-1:
                return nums[-1]
            else:
                cm = max(nums[index]+helper(index+2), helper(index+1))
                return cm
        return helper(0)
